More Crypto, SSL server, crt/key as Resource, HashingContext

Faless opened the pull request:
This PR includes:

- New `hex_encode` function for `PoolByteArray` that convert the bytes to an hex string.
- New **`Crypto`** class that can **generate random bytes, rsa keys,  self signed certificates**.
- Certificates (**crt**) and Keys (**key**) are considered **`Resource`**. You can `load` and save them.
- Implemented `StreamPeerSSL.accept_stream` (**SSL server**) and allow selecting specific valid CAs when calling `conntect_to_stream`.
- ~**AES encryption** via the `AESContext` class~ (will be done in a separate PR).
- **MD5/SHA1/SHA256 hashing** via the `HashingContext` class.

This is also some ground work towards implementing DTLS, and internal HTTP(S) server for HTML5 debug.

Some interesting points:
- Certs and keys gets "locked" when in use by a stream, and can't be changed when in use.
  This is so you can have a single instance of key/cert, and use it across multiple peers, without holding.
- Should we have a default "Crypto" singleton (right now it's an object you instance)? I like having the possibility of them having different seeds, but having a default one would be nice.
- Should we use the same random number generator across all SSL sessions (could that be a weakness?), right now it uses one rng per session as before.

A small example that uses this PR (SSL server with self-signed certificate, visit https://localhost:4343 after opening):
```gdscript
extends Node

class Client:
	var ssl = StreamPeerSSL.new()
	var recv = ""

	func set_stream(stream, key, cert):
		ssl.blocking_handshake = false
		ssl.accept_stream(stream, key, cert)

	func process():
		if ssl.get_status() != StreamPeerSSL.STATUS_CONNECTED:
			return
		ssl.poll()
		if ssl.get_available_bytes() > 0:
			recv += ssl.get_data(ssl.get_available_bytes())[1].get_string_from_utf8()
		if recv.ends_with("\r\n\r\n"):
			ssl.put_data(("HTTP/1.0 200 OK\r\nContent-Type: text/html\r\n\r\n" + \
						"<h2>Godot TLS Test Server</h2>\r\n"  + \
						"<p>Successful connection using SSL</p>\r\n").to_utf8())
			ssl.disconnect_from_stream()

var _server = TCP_Server.new()
var _clients = []
var _cert = null
var _key = null

func _ready():
	var crypto = Crypto.new()
	_key = crypto.generate_rsa(4096)
	_cert = crypto.generate_self_signed_certificate(_key, "CN=aserver,O=A Game Company,C=IT")
	_key.save("res://generated.key")
	_cert.save("res://generated.crt")
	_server.listen(4343)

func _process(delta):
	if _server.is_connection_available():
		var c = Client.new()
		c.set_stream(_server.take_connection(), _key, _cert)
		_clients.append(c)

	for c in _clients:
		c.process()
```

Another example showing hahsing:
```gdscript
extends Node

func _ready():
	var ctx = HashingContext.new()
	ctx.start(HashingContext.HASH_MD5) # Or HASH_SHA1 or HASH_SHA256
	print("a".to_utf8().size())
	ctx.update("a".to_ascii())
	var fin = ctx.finish()
	print(Array(fin))
	print(fin.hex_encode())
```

Fixes #27045 .

References (possibly close after some more work): #29871 #25928 #1969

nate066 commented:
This looks really cool, I hope AES support will happen along with a hashing algorithm like argon2. :)

fire commented:
How do you want this reviewed? Is there anything in particular you want to look for? Did you make basic demos that use these and do they run as expected?

fire commented:
Also, can you improve the quality of the commity messages and reduce the number of commits?

Faless commented:
> How do you want this reviewed?

It's still WIP, so mainly in terms of structure (see interesting points), 

>  Did you make basic demos that use these and do they run as expected?

Just the attached script.

> Also, can you improve the quality of the commit messages and reduce the number of commits?

Yes, I will, before removing the WIP tag, but it's likely going to be 2 commits in the end.
- Adding crypto
  (`modules/mbdetls/crypto_mbedtls.*` and `core/math/crypto.*`)
- Porting StreamPeerSSL to crypto
  (`core/io/stream_peer_ssl.*`, `modules/mbdetls/ssl_context_mbed_tls.*`, `modules/mbdetls/stream_peer_mbedtls.*`)

aspenforest commented:
Please consider adding SHA-3 and BLAKE hash functions. :)

akien-mga commented:
FYI, the javascript platform currently disables the mbedTLS module with this comment:
```python
        # Disabling the mbedtls module reduces file size.
        # The module has little use due to the limited networking functionality
        # in this platform. For the available networking methods, the browser
        # manages TLS.
        ('module_mbedtls_enabled', False),
```
I guess we might want to enable it again for this module?

Faless commented:
> I guess we might want to enable it again for this module?

Yeah, an alternative is for Javascript to have its own implementation via the [CryptoAPI](https://developer.mozilla.org/en-US/docs/Web/API/Crypto).

But I would prefer having only one implementation for ease of maintenance, it should pretty much work out of the box, the only difference might be in the entropy function.

I'll do a size comparison with and without the module to see how much bigger we get in the HTML5 export.

fire commented:
Are you blocked on anything? Is there anything we can do to help you?

Faless commented:
> Are you blocked on anything? Is there anything we can do to help you?

I've been busy on the WebSocket library switch and some other fixes.
I'm finally back to this, so you can expect an update soon, I'm thinking of adding AES and support to streamed hashing before marking this as ready for review :)

Faless committed:
Fix CryptoCore signatures, add SHA1 context.

Fix hash size in SHA256 signature
Fix source parameter in hash context update function to be const.
Add SHA1 hash context.

Faless committed:
Move CryptoCore to it's own folder.

Crypto classes will be placed in core/crypto.

Faless committed:
Add Crypto custom class

fire commented:
Can you split those changes to another PR, or do they require code stubs for them? 

I wish to try this in master without custom merging this branch.

Faless committed:
Add hex_encode function to PoolByteArray

Faless committed:
Add hashing context class

Faless committed:
New CryptoMbedTLS Crypto implementation.

Allows random bytes, RSA keys, and X509 certificates generation.

Faless committed:
Rewrite StreamPeerSSL with SSLContext helper class

connect_to_stream now accepts optional parameter to specify which
certificates to trust.
Implement accept_stream (SSL server) with key/cert parameters to specify
the RSA key and X509 certificate resources.

nate066 commented:
Is there any chance you add CBC or GCM to AES. Since ECB is considered less secure/ insecure. https://datalocker.com/what-is-the-difference-between-ecb-mode-versus-cbc-mode-aes-encryption/

Faless commented:
> Is there any chance you add CBC or GCM to AES.

Of course, but likely in a separate PR, this one is already pretty big.

nate066 commented:
That would be awesome

Faless commented:
I've removed the AES part, since it was incomplete and it was taking too long for me to get it right.
AES will be added in a separate PR, this one should be ready now.

akien-mga commented on editor/editor_settings.cpp:
```diff
@@ -1022,7 +1022,8 @@ void EditorSettings::setup_network() {
 
 	// Editor SSL certificates override
-	_initial_set("network/ssl/editor_ssl_certificates", _SYSTEM_CERTS_PATH);
+	String certs = has_setting("network/ssl/editor_ssl_certificates") ? get("network/ssl/editor_ssl_certificates") : _SYSTEM_CERTS_PATH;
```
Why is it done this way? Doesn't it mean that the "initial" (revert) value will always be the currently set one, so it's not revertable?

Faless replied:
No, it means that the value shown in the editor settings will be the last configured value if available, otherwise `_SYSTEM_CERTS_PATH` (which is an empty string).
The same is done above for debug address and port.

akien-mga replied:
But isn't it what `_initial_set` did already? It's designed so that it defines the default value, but doesn't override any custom value.

Faless replied:
I see, I'll fix that, probably `network/debug/remote_port` should use the same behaviour then. (`network/debug/remote_host` is a bit special instead)

Faless replied:
Well, `_initial_set` always calls `set` :( . I guess `setup_network` needs to be modified a bit.

akien-mga replied:
I guess `setup_network()` simply happens too late? `_initial_set` works fine for dozens of settings, but it's meant to be called before loading the editor settings. It sets the default value, and then it can be overridden by user values from `editor_settings-3.tres`.

akien-mga replied:
See `_load_defaults()`, which is called by the constructor.

Faless replied:
See added commit https://github.com/godotengine/godot/pull/29871/commits/fc411115f5ddb214c19b23499400f10266095e5f as a potential fix

akien-mga replied:
Looks good!

akien-mga commented on core/crypto/crypto.cpp:
```diff
@@ -0,0 +1,173 @@
+}
+
+PoolByteArray Crypto::generate_random_bytes(int p_bytes) {
+	ERR_EXPLAIN("generate_random_bytes is not available when mbetls module is disabled");
```
See #31207, this should be changed to `ERR_FAIL_V_MSG`.

There's also a typo in "mbe**d**tls" (in all occurrences), and it would be nice to add a dot at the end of the error messages.

Faless replied:
Thanks, fixed!

Faless committed:
Fix Network Editor Settings setup.

akien-mga commented:
Thanks!
